### PR TITLE
substrate tx queue

### LINKED ISSUE #191: [Task] Implement Transaction queue for Substrate
TxQueue stores transaction requests, so the relayer can process them later, this prevents issues such as creating transactions with the same nonce.

## Task Checklist
- [x] Implement transaction queue for substrate
- [x] Update `SignatureBridgeWacther` to process proposals execution through Tx Queue

----
Starts the TxQueue service.
Returns a future that resolves `Ok(())` on success, otherwise returns an error.

Examples
```
use crate::tx_queue::TxQueue;
let tx_queue = TxQueue::new(ctx, chain_name.clone(), store);
let task = async move {
    tokio::select! {
     _ = tx_queue.run() => {
         // do something
          },
    _ = shutdown_signal.recv() => {
         // do something
         },
    }
};
```

TxQueue implementation for [Evm](https://github.com/webb-tools/relayer/blob/main/src/tx_queue.rs)

### PR DESCRIPTION
**Pull request by salman01zp**
## Transaction Queue for substrate
- Enqueue transaction call-data (pallet u8, call u8, call params u8)  in `TransactionQueue: Vec<u8>`
- Dequeue tx call_data and construct extrinsic as implemented [here](https://docs.rs/subxt/latest/src/subxt/client.rs.html#424-445)

```rust
// for dkg-substrate. It uses PolkadotExtrinsicParams (A struct representing the signed extra and additional parameters 
//required to construct a transaction for a polkadot node.)
let tx_queue = SubstrateTxQueue::new(ctx, chain_id, store);
tx_queue.run::<subxt::PolkadotExtrinsicParams<subxt::DefaultConfig>>()


// for protocol-substrate. It uses SubstrateExtrinsicParams (A struct representing the signed extra and additional parameters 
//required to construct a transaction for the default substrate node.)
let tx_queue = SubstrateTxQueue::new(ctx, chain_id, store);
tx_queue.run::<subxt::SubstrateExtrinsicParams<subxt::DefaultConfig>>()
```

       
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### Reference issue to close (if applicable)
- Closes #191 

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by salman01zp** [comment]
blocked by #231. Substrate tests are failing due to changes in protocol-substrate

**Comment by salman01zp** [comment]
This subxt [PR](https://github.com/paritytech/subxt/pull/593) should make it possible to build, store and later submit calls more easier

File: src/events_watcher/substrate/signature_bridge_watcher.rs [hunk 1]
```diff
@@ -177,87 +172,43 @@ where
...
+        // call data bytes (pallet u8, call u8, call params).
+        let call_data = {
+            let mut bytes = Vec::new();
+            let locked_metadata = api.client.metadata();
+            let metadata = locked_metadata.read();
+            let pallet = metadata.pallet(ExecuteProposal::PALLET)?;
+            bytes.push(pallet.index());
+            execute_proposal_call.encode_to(&mut bytes);
+            bytes
+        };
```
[the remark this answers is not in the document]
Comment (shekohex) [reply src/events_watcher/substrate/signature_bridge_watcher.rs:192]: Here is an example: 
```rs
fn encode_call_data<C: Call>(metadata: Arc<RwLock<Metadata>>, call: C) -> Vec<u8> {
    let mut bytes = Vec::new();
    let metadata = metadata.read();
    let pallet = metadata.pallet(C::PALLET)?;
    bytes.push(pallet.index());
    call.encode_to(&mut bytes);
    bytes
}
```

File: src/events_watcher/substrate/signature_bridge_watcher.rs [hunk 2]
```diff
@@ -307,63 +258,33 @@ where
...
+        // call data bytes (pallet u8, call u8, call params).
+        let call_data = {
+            let mut bytes = Vec::new();
+            let locked_metadata = api.client.metadata();
+            let metadata = locked_metadata.read();
+            let pallet = metadata.pallet(SetMaintainer::PALLET)?;
+            bytes.push(pallet.index());
+            set_maintainer_call.encode_to(&mut bytes);
+            bytes
+        };
```
Comment (shekohex) [inline src/events_watcher/substrate/signature_bridge_watcher.rs:276]: Same here, using a reusable function would make sense and reduce errors.

File: src/tx_queue/substrate_tx_queue.rs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+                // dequeue transaction call data. This are call params stored as bytes
+                let maybe_call_data = store.dequeue_item(
+                    SledQueueKey::from_substrate_chain_id(chain_id),
+                )?;
+                if let Some(call_data) = maybe_call_data {
+                    let call_data = subxt::Encoded(call_data);
+                    tracing::debug!(
+                        "Transaction call in hex : {:?}",
+                        hex::encode(&call_data.encode())
+                    );
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:145]:
```suggestion
                    tracing::trace!(
                        "Transaction call in hex : {:?}",
                        hex::encode(&call_data.encode())
                    );
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+                        match e {
+                            TransactionStatus::Future => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Future",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:251]:
```suggestion
                                    status = "Future",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::Ready => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Ready",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:262]:
```suggestion
                                    status = "Ready",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::Broadcast(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Broadcast",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:273]:
```suggestion
                                    status = "Broadcast",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::InBlock(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "InBlock",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:284]:
```suggestion
                                    status = "InBlock",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::Retracted(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Retracted",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:295]:
```suggestion
                                    status = "Retracted",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::FinalityTimeout(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "FinalityTimeout",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:306]:
```suggestion
                                    status = "FinalityTimeout",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+                            TransactionStatus::Finalized(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Finalized",
+                                    finalized = true,
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:318]:
```suggestion
                                    finalized = true,
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+                            TransactionStatus::Usurped(_) => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Usurped",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:331]:
```suggestion
                                    status = "Usurped",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::Dropped => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Dropped",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:342]:
```suggestion
                                    status = "Dropped",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+                            }
+                            TransactionStatus::Invalid => {
+                                tracing::event!(
+                                    target: crate::probe::TARGET,
+                                    tracing::Level::DEBUG,
+                                    kind = %crate::probe::Kind::TxQueue,
+                                    ty = "SUBSTRATE",
+                                    chain_id = %chain_id.as_u64(),
+                                    status = "Invalid",
+
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:353]:
```suggestion
                                    status = "Invalid",
```

File: src/tx_queue/substrate_tx_queue.rs [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+        // get pair
+        let pair = self.ctx.substrate_wallet(&chain_id.to_string()).await?;
+        let signer: PairSigner<subxt::DefaultConfig, Pair> =
+            PairSigner::new(pair);
+
+        tracing::event!(
+            target: crate::probe::TARGET,
+            tracing::Level::DEBUG,
+            kind = %crate::probe::Kind::TxQueue,
+            ty = "Substrate",
```
Comment (shekohex) [inline src/tx_queue/substrate_tx_queue.rs:128]:
```suggestion
            ty = "SUBSTRATE",
```

**Review by shekohex — CHANGES_REQUESTED** [review]
Looks good, I've left few comments for improvements.

File: src/tx_queue/substrate/call_data.rs [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+// call data bytes (pallet u8, call u8, call params).
+pub fn encode_call_data<C: Call>(
+    metadata: Arc<RwLock<Metadata>>,
+    call: C,
+) -> anyhow::Result<Vec<u8>> {
+    let mut bytes = Vec::new();
+    let metadata = metadata.read();
+    let pallet = metadata.pallet(C::PALLET)?;
+    bytes.push(pallet.index());
+    bytes.push(pallet.call_index::<C>()?);
```
Comment (shekohex) [inline src/tx_queue/substrate/call_data.rs:27]: Is the call index is now needed? I guess you mentioned before that the call contains already the call index?

Comment (salman01zp) [reply src/tx_queue/substrate/call_data.rs:27]: yeah. I changed its type from
(This does not implement the `Call` trait, This struct on encoding includes call index )
`webb::substrate::protocol_substrate_runtime::api::runtime_types::pallet_signature_bridge::pallet::Call`

To (this struct implements Call trait)
`webb::substrate::protocol_substrate_runtime::api::signature_bridge::calls::ExecuteProposal`

**Review by shekohex — APPROVED** [review]
LGTM!

Good work, Salman.

**Pull request merged**